src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributes.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TagHelperAttributes : ReadOnlyTagHelperAttributes<TagHelperAttribute>, IList<TagHelperAttribute>
+    {
+        public new TagHelperAttribute this[int index]
+        {
+            get
+            {
+                return base[index];
+            }
+            [param: NotNull]
+            set
+            {
+                _attributes[index] = value;
+            }
+        }
+
+        public new TagHelperAttribute this[[NotNull] string key]
+        {
+            get
+            {
+                return base[key];
+            }
+            [param: NotNull]
+            set
+            {
+                var attributeReplaced = false;
+
+                for (var i = _attributes.Count - 1; i >= 0; i++)
+                {
+                    if (SameKey(key, _attributes[i]))
+                    {
+                        // We replace the last attribute with the provided value, remove all the rest.
+                        if (!attributeReplaced)
+                        {
+                            attributeReplaced = true;
+                            _attributes[i] = value;
+                        }
+                        else
+                        {
+                            _attributes.RemoveAt(i);
+                        }
+                    }
+                }
```

Comment:
if `!attributeReplaced` at this point, need to add a new entry to the list.  would be very unexpected for the setter to do nothing when there's no existing entry with same name.